base/compiler/tfuncs.jl

```diff
@@ -1474,7 +1474,7 @@ add_tfunc(fieldtype, 2, 3, fieldtype_tfunc, 0)
 # Like `valid_tparam`, but in the type domain.
 valid_tparam_type(T::DataType) = valid_typeof_tparam(T)
 valid_tparam_type(U::Union) = valid_tparam_type(U.a) && valid_tparam_type(U.b)
-valid_tparam_type(U::UnionAll) = valid_tparam_type(unwrap_unionall(U))
+valid_tparam_type(U::UnionAll) = false
```

Comment:
On further reflection, I'm not sure I like this. It is true that `valid_typeof_tparam` only handles concrete types, but it could certainly handle all types with only concrete subtypes, which we do have a query for, in which case this is correct.

Reply:
I guess your thinking is that eventually `valid_typeof_tparam` should assume that all of the UnionAll bounds were present and thus must get assigned, so that at runtime it must turn into a valid concrete object, whose properties we can pre-determine there? That sounds reasonable to me.

Reply:
should I close this then?

Reply:
> I guess your thinking is that eventually `valid_typeof_tparam` should assume that all of the UnionAll bounds were present and thus must get assigned, so that at runtime it must turn into a valid concrete object, whose properties we can pre-determine there? That sounds reasonable to me.

Well, this is a lattice query, so the question is whether all of the runtime values that correspond to it have `valid_tparam` being true. I think that should be true for things like `struct Foo{T}; end` where we know that all runtime values that are `::Foo` will be valid tparams even if we don't know what `T` is.